core/ir/synthetic/decode.c

```diff
@@ -80,26 +85,29 @@ decode_from_synth(dcontext_t *dcontext, byte *encoded_instr, instr_t *instr)
 
     /* Decode synthetic opcode as instruction category.
      */
-    uint category = (*encoding & CATEGORY_MASK) >> CATEGORY_SHIFT;
+    uint category = (encoding & CATEGORY_MASK) >> CATEGORY_SHIFT;
     instr_set_category(instr, category);
 
-    /* Decode register destination operands, if present.
+    /* Decode register destination operands and their sizes, if present.
      */
-    for (uint i = 0; i < num_dsts; ++i) {
-        reg_id_t dst = (reg_id_t)encoded_instr[i + INSTRUCTION_BYTES];
-        // TODO i#6662: need to add virtual registers.
-        // Right now using regular reg_id_t (which holds DR_REG_ values) from opnd_api.h.
+    uint num_dst_bytes = num_dsts * OPERAND_BYTES;
+    for (uint i = 0; i < num_dst_bytes; i += OPERAND_BYTES) {
+        reg_id_t dst = (reg_id_t)encoded_instr[i + HEADER_BYTES];
+        opnd_size_t dst_size = (opnd_size_t)encoded_instr[i + 1 + HEADER_BYTES];
```

Comment:
We had agreed to have a single operation size for the whole instruction, not per-operand.

Reply:
Fixed.